src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +340,477 @@ where
+                Err(_recv_error) => {
+                    // Sender hung up, this seems fine and can happen when
+                    // shutting down.
+                    notify_listeners(responders, || {
+                        Err(StorageError::ShuttingDown("PersistMonotonicWriteWorker"))
+                    });
+
+                    // End the task since we can no longer send writes to persist.
+                    return ControlFlow::Break("sender hung up".to_string());
+                }
+            };
+
+            match append_result {
+                // Everything was successful!
+                Ok(()) => {
+                    // Notify all of our listeners.
+                    notify_listeners(responders, || Ok(()));
+
+                    self.current_upper = new_upper;
+
+                    // Very important! This is empty at steady state, while
+                    // desired keeps an in-memory copy of desired state.
+                    self.to_write.clear();
+
+                    tracing::debug!(%self.id, "appended to differential collection");
+
+                    // Break out of the retry loop so we can wait for more data.
+                    break;
+                }
+                // Failed to write to some collections,
+                Err(StorageError::InvalidUppers(failed_ids)) => {
+                    // Someone else wrote to the collection. We need to read
+                    // from persist and update to_write based on that and the
+                    // desired state.
+
+                    assert_eq!(
+                        failed_ids.len(),
+                        1,
+                        "received errors for more than one collection"
+                    );
+                    assert_eq!(
+                        failed_ids[0].id, self.id,
+                        "received errors for a different collection"
+                    );
+
+                    let actual_upper = if let Some(ts) = failed_ids[0].current_upper.as_option() {
+                        ts.clone()
+                    } else {
+                        return ControlFlow::Break("upper is the empty antichain".to_string());
+                    };
+
+                    tracing::info!(%self.id, ?actual_upper, expected_upper = ?self.current_upper, "retrying append for differential collection");
+
+                    // We've exhausted all of our retries, notify listeners and
+                    // break out of the retry loop so we can wait for more data.
+                    if retries.next().await.is_none() {
+                        notify_listeners(responders, || {
+                            Err(StorageError::InvalidUppers(failed_ids.clone()))
+                        });
+                        error!(
+                            "exhausted retries when appending to managed collection {failed_ids:?}"
+                        );
+                        break;
+                    }
```

Comment:
Does not seem to be covered by tests according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

Reply:
We can't currently trigger this code path because we don't have concurrent writers. We might have that in the future.